airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java

```diff
@@ -28,63 +36,131 @@ public class AutoDisableConnectionActivityImpl implements AutoDisableConnectionA
-        // jobs), do not auto-disable
         if (numFailures == 0) {
           return new AutoDisableConnectionOutput(false);
+        } else if (numFailures == configs.getMaxFailedJobsInARowBeforeConnectionDisable()) {
+          // disable connection if max consecutive failed jobs limit has been hit
+          disableConnection(input.getConnectionId(), lastJob);
+          return new AutoDisableConnectionOutput(true);
+        } else if (numFailures == maxFailedJobsInARowBeforeConnectionDisableWarning) {
+          // warn if number of consecutive failures hits 50% of MaxFailedJobsInARow
+          jobNotifier.autoDisableConnectionAlertWithoutCustomerioConfig(CONNECTION_DISABLED_WARNING_NOTIFICATION, lastJob);
+          return new AutoDisableConnectionOutput(false);
         }
 
-        // if the very first job of a connection fails, it will hit the condition of "only failed jobs in
-        // the past `maxDaysOfOnlyFailedJobs` days", to avoid this behavior, we ensure that this condition
-        // is only taken into account if the connection has a job that's at least `maxDaysOfOnlyFailedJobs`
-        // days old
-        if (numFailures != configs.getMaxFailedJobsInARowBeforeConnectionDisable()) {
-          final Optional<Job> optionalFirstJob = jobPersistence.getFirstReplicationJob(input.getConnectionId());
-          if (optionalFirstJob.isPresent()) {
-            final long timeBetweenCurrTimestampAndFirstJob = input.getCurrTimestamp().getEpochSecond()
-                - optionalFirstJob.get().getCreatedAtInSecond();
-            if (timeBetweenCurrTimestampAndFirstJob <= TimeUnit.DAYS.toSeconds(maxDaysOfOnlyFailedJobs)) {
-              return new AutoDisableConnectionOutput(false);
-            }
-          }
+        // calculate the number of days this connection first tried a replication job, used to ensure not to
+        // disable or warn for `maxDaysOfOnlyFailedJobs` if the first job is younger than
+        // `maxDaysOfOnlyFailedJobs` days, This avoids cases such as "the very first job run was a failure".
+        final Job firstJob = jobPersistence.getFirstReplicationJob(input.getConnectionId())
+            .orElseThrow(() -> new Exception("Auto-Disable Connection should not have been attempted if no replication job has been run."));
+        final int numDaysSinceFirstReplicationJob = getDaysSinceTimestamp(currTimestampInSeconds, firstJob.getCreatedAtInSecond());
+
+        // disable connection if only failed jobs in the past maxDaysOfOnlyFailedJobs days
+        if (numDaysSinceFirstReplicationJob >= maxDaysOfOnlyFailedJobs && successTimestamp.isEmpty()) {
+          disableConnection(input.getConnectionId(), lastJob);
+          return new AutoDisableConnectionOutput(true);
+        }
+
+        // sending warning for disable connection if only failed jobs in the past
+        // maxDaysOfOnlyFailedJobsBeforeWarning days
+        if (numDaysSinceFirstReplicationJob >= maxDaysOfOnlyFailedJobsBeforeWarning &&
```

Comment:
This is a bit hard to follow because of the nesting, can you pull some of the inner conditional logic out into named variables to help with readability?

Something like:
```
final boolean firstReplicationIsOldEnough = numDaysSinceFirstReplicationJob >= maxDaysOfOnlyFailedJobsBeforeWarning;
final boolean enoughDaysPassedSinceLastSuccess = successTimestamp.isEmpty()
                || getDaysSinceTimestamp(currTimestampInSeconds, successTimestamp.get()) >= maxDaysOfOnlyFailedJobsBeforeWarning);

if (firstReplicationIsOldEnough && enoughDaysPassedSinceLastSuccess) {
    ...
}
```

You should pick better variable names if you can think of them, especially if mine don't actually match the logic 😅

Reply:
Thanks for the suggestions. I edited the above to help with readability.

I went with uglier, longer names like `firstReplicationOlderThanMaxDisableWarningDays` and `successOlderThanPrevFailureByMaxWarningDays` to help make things as clear as possible, but do let me know if they actually make things less clear, happy to go with your suggestions if you think so.